### PR TITLE
Add limits data for all xija model MSIDs in MCC thermal characteristics

### PR DESCRIPTION
**Pull request by taldcroft**
This adds `limits` specifications for each of the 11 thermal models that have limits in the MATLAB `characteristics_Thermal.m` file.  See below.

This adds fields of the form `<limit>.matlab_identifier` to document the mapping to the MATLAB characteristics and potentially allow limits to be set dynamically. This draws on the template `<limit_system>.<limit_type>.<high/low>.<optional qualifier>` defined in https://github.com/sot/xija/issues/70, where the `<optional qualifier>` is the MATLAB identifier.

I did my best to do the mapping from what is in the characteristics to the https://github.com/sot/xija/issues/70 spec, but this needs review.  The `minusyz` model spec has a lot of diffs from trailing whitespace that was cleaned up. These whitespace diffs can be hidden on the diff page by hitting the Settings icon and selecting the checkbutton to ignore whitespace.

```
Thermal.msid_pm1thv2t.yellow_limit    = 210;  # planning hi
Thermal.msid_pm1thv2t.red_limit       = 240;

Thermal.msid_pm2thv1t.yellow_limit    = 210;  # planning hi
Thermal.msid_pm2thv1t.red_limit       = 240;

Thermal.msid_pline03t.yellow_limit    = 200;   % degF  # planning hi
Thermal.msid_pline03t.red_limit       = -10000; % degF
Thermal.msid_pline03t.reference_limit = 50;   % degF  # planning low

Thermal.msid_pline04t.yellow_limit    = 170;    % degF  # planning hi
Thermal.msid_pline04t.red_limit       = -10000; % degF
Thermal.msid_pline04t.reference_limit = 50;     % degF  # planning low

Thermal.msid_tcylaft6.yellow_limit    = 130; % degF  # planning hi
Thermal.msid_tcylaft6.red_limit       = -10000;
Thermal.msid_tcylaft6.reference_limit = -10000; % degF

Thermal.msid_1dpamzt.yellow_limit    = 39.5;   % degC
Thermal.msid_1dpamzt.red_limit       = 41.5; % degC
Thermal.msid_1dpamzt.reference_limit = 37.5;   % degC  # planning hi
Thermal.msid_1dpamzt.low_limit       = 12; % deg C  # planning low

Thermal.msid_1deamzt.yellow_limit    = 39.5;   % degC
Thermal.msid_1deamzt.red_limit       = 42.5; % degC
Thermal.msid_1deamzt.reference_limit = 37.5;   % degC  # planning hi

Thermal.msid_pftank2t.yellow_limit    = 120; % degF
Thermal.msid_pftank2t.red_limit       = 140; % degF
Thermal.msid_pftank2t.reference_limit = 110;  % degF  # planning hi

Thermal.msid_aacccdpt_c.yellow_limit    = -10000; % degC
Thermal.msid_aacccdpt_c.red_limit       = -10000; % degC
Thermal.msid_aacccdpt_c.reference_limit = -6.5; % degC  # planning hi
Thermal.msid_aacccdpt_c.penalty_limit   = -7.5; % degC  # penalty limit

Thermal.msid_4rt700t.yellow_limit    = 110; % degF
Thermal.msid_4rt700t.red_limit       = 120; % degF
Thermal.msid_4rt700t.reference_limit = 100; % degF  # planning hi

Thermal.msid_1pdeaat.yellow_limit    = 57; % degC
Thermal.msid_1pdeaat.red_limit       = 62; % degC
Thermal.msid_1pdeaat.reference_limit = 57 - 4.5; % degC  # planning hi
```

### CODE REVIEW

**Review by jzuhone — APPROVED** [review]

**Comment by taldcroft** [comment]
It doesn't have to be done in this PR, but the ACIS FP model spec is not updated here. It is complicated and I don't actually know how all the limits are done in `xija_gui_fit` and MATLAB tools.

File: chandra_models/xija/pftank2t/pftank2t_spec.json
Comment (jskrist) [inline chandra_models/xija/pftank2t/pftank2t_spec.json:154]: This limit was just updated in the MATLAB dev branch (Trunk) to be `115` I'm not sure when this change was expected to go live, but it should probably be updated here too.

File: chandra_models/xija/pline/pline03t_model_spec.json
Comment (jskrist) [inline chandra_models/xija/pline/pline03t_model_spec.json:162]: similar to the MUPS models, should this be `planning.caution.high` as a yellow limit?

File: chandra_models/xija/pline/pline04t_model_spec.json
Comment (jskrist) [inline chandra_models/xija/pline/pline04t_model_spec.json:202]: same as PLINE03T

File: chandra_models/xija/minusyz/minusyz_spec.json
Comment (jskrist) [inline chandra_models/xija/minusyz/minusyz_spec.json:323]: the MATLAB tools use the file `tcylaft6_spec.json` for this model.  It looks like they don't use this file at all.  Should the limits for tcylaft6 be added to it's spec file directly?

Also same question as in other places about naming convention.  Should this be a `caution` instead of a `warning`?

File: chandra_models/xija/pftank2t/pftank2t_spec.json
Comment (jskrist) [inline chandra_models/xija/pftank2t/pftank2t_spec.json:157]: the MATLAB tools characteristics are also storing `pftankip` (tank pressure limits).  I don't think the xija models use these values, but wasn't sure if we wanted to store all associated limits in one place by adding them here.

**Review by jskrist — COMMENTED** [review]
I may be confused about the distinction between `warning` and `caution` limits, so please correct me if I am.

I noticed that the only limit characteristics not covered by this update are the values for the ACIS_FP model.  These characteristics don't fit the typical format used here, but I would put them under the category of 'limits we may want to change without a full MATLAB release that relate to the thermal models'  Should these characteristics values be captured in here?

**Comment by jzuhone** [comment]
> It doesn't have to be done in this PR, but the ACIS FP model spec is not updated here. It is complicated and I don't actually know how all the limits are done in `xija_gui_fit` and MATLAB tools.

@taldcroft I have an update to the ACIS FP model coming up so I'll include it there.

**Comment by jzuhone** [comment]
@jskrist "planning.warning" means we can plan to this limit but not cross it. "odb.caution" is typically referred to as a "yellow" limit, and "odb.warning" is typically referred to as a "red" limit.

File: chandra_models/xija/dpa/dpa_spec.json
Comment (jzuhone) [reply chandra_models/xija/dpa/dpa_spec.json:270]: Actually @taldcroft we fail a load if it goes below the +12 C limit, so this should probably be `planning.warning.low`.

Comment (matthewdahmer) [reply chandra_models/xija/dpa/dpa_spec.json:270]: I was under the impression we fail the load under 12C under specific circumstances. We can be under 12C if 3 FEPs are on for example.

**Comment by matthewdahmer** [comment]
In order to reduce confusion, it would be simpler to only define `planning.warning.<low, high>` and `planning.caution.<low, high>` if these values are defined in a guideline. For example, the ACA has both planning and penalty limits, these seem to map to `planning.warning.high` and `planning.caution.high` fairly well, however the PLINE models do not have any maximum temperature limits, so no high planning limits should be defined.

File: chandra_models/xija/mups_valve/pm1thv2t_spec.json
Comment (taldcroft) [reply chandra_models/xija/mups_valve/pm1thv2t_spec.json:131]: The MUPS guideline says "Do not exceed 210 deg F for the predicted MUPS thruster valve temperatures ( PM1THV2T, or PM2THV1T) as determined by the MCC thermal model predictions". So 210 is the official do-not-exceed planning limit => `planning.warning.high`. That planning limit is apparently `yellow_limit` in MATLAB.

The 240 limit is the ODB hardware limit. (I'm not 100% sure it is in the ODB, but that value is the don't-ever-exceed-or-it-might explode limit. In our nomenclature that would be `odb.warning.high`. 

I've been a little liberal with `odb`, but maybe we should call that instead `hardware` to imply a real hardware limit regardless of where it might be captured.

File: chandra_models/xija/pftank2t/pftank2t_spec.json
Comment (taldcroft) [reply chandra_models/xija/pftank2t/pftank2t_spec.json:157]: I suppose we could make a `pftankip_spec.json` model here that it not actually a xija model but has the `limits` data structure. That's slightly odd but would make it possible to change tank pressure planning limits in the same way. Fine by me.

File: chandra_models/xija/pline/pline03t_model_spec.json
Comment (taldcroft) [reply chandra_models/xija/pline/pline03t_model_spec.json:162]: Mostly the MATLAB `yellow_limit` corresponds to the planning limit (`planning.warning.high`) from what I could see.

File: chandra_models/xija/dea/dea_spec.json
Comment (taldcroft) [inline chandra_models/xija/dea/dea_spec.json:263]: The ACIS models are oddballs in labeling the planning limit as `reference_limit`.

**Review by taldcroft — COMMENTED** [review]
There seems to be a mixture of conventions. ACA and ACIS seem to use `reference_limit` at the MATLAB identifier for the planning limit, while most (all?) of the others use `yellow_limit`.  In the spec files I'm consistently using `planning.warning.high/low` as the hard MP planning limits.

**Comment by taldcroft** [comment]
BTW I annotated the limits in the top description to show an overview of the current naming situation in the MATLAB characteristics.

File: chandra_models/xija/mups_valve/pm1thv2t_spec.json
Comment (jskrist) [reply chandra_models/xija/mups_valve/pm1thv2t_spec.json:131]: I like the `hardware` & `planning` naming convention (vs. `odb` & `planning`).  I think it is a bit more intuitive/explicit.

File: chandra_models/xija/pftank2t/pftank2t_spec.json
Comment (jskrist) [reply chandra_models/xija/pftank2t/pftank2t_spec.json:157]: If these values aren't really being used/validated in the SOT tools at the moment, I'd prefer that they stay in the MATLAB Characteristics, where any changes will get some validation from the existing regression testing.  If they are used/tested in the SOT release workflow then I am happy either way.

Comment (taldcroft) [reply chandra_models/xija/pftank2t/pftank2t_spec.json:157]: Tank pressure limits are not in the SOT tools to my knowledge. So 👍 to keeping them in the characterisitics.

**Comment by matthewdahmer** [comment]
Why are we defining Matlab identifiers in the model json files? It seems as if this should be defined in the Matlab characteristics file since it affects how the Matlab code functions, has no bearing on how any Python code functions (and if it does it should be isolated to that specific code base), and seems to only be causing unnecessary confusion. From another perspective, if the Matlab code changes how it interprets a limit, the only changes necessary should be in Matlab code.

**Review by matthewdahmer — CHANGES_REQUESTED** [review]
It seems as if the conversations referencing specific code segments in the "files changed" tab seemed to have run their course so I marked them as resolved, however there are two limit changes that are necessary before merging and I'd like to work out why we need Matlab identifiers in these files. Here are my notes for reference:

**Notes that do not reference required changes:**

**`odb` vs `hardware` terminology**
I prefer the curently used `odb` terminology. The purpose of the `odb` (or `tdb` as a subset of the `odb`) is to define the safety limits. We tend to use `hardware` limits interchangeably with `safety` and `odb` limits, so I see the appeal here. That being said, using a more human readable term obscures the source and could be a source of confusion for someone who is less familiar with these limits (e.g. is there a difference between "hardware" and "odb" limits?). Along those lines, using `guideline` instead of `planning` also seems more appropriate for such limits, however we are already using `planning` for the approved aca model so I'm ok with sticking with `planning`.

**When to include limits**
As is currently the case, only those limits that are used should be included. This reduces required model maintenance, so these models do not need to be updated when unused limits are changed.

**Hardcoded values**
It seems as if we are in agreement that there are circumstances where some limits can be left as hardcoded values in the Matlab characteristics file (e.g. PFTANKIP). 

**Minusyz model**
This model is no longer used, is grossly innacurate, and should be deleted, though not necessarily with this release. I bring this up here since it was modified to include tcylaft6 limits. Is the `minusyz_spec.json` model used in testing anywhere? I seem to remember seeing some testing output that referenced this model.

**tcylaft6 model**
We are not using this model to limit central cylinder temperatures anymore, and if we do find we need it, we can always re-implement it. I vote for deleting this model from `chandra_models` once it is convenient to do so from the Matlab FOT Tools perspective, though not with this release. This also means we don't need to add any limits to this file.

**Notes that reference required changes, or possible changes:**

**Limit corrections:**
The `planning.warning.high` PFTANK2T limit should be 115F
The `planning.warning.high` 4RT700T limit should be 103F

**Matlab Identifiers**
Why are we defining Matlab identifiers in the model json files? It seems as if this should be defined in the Matlab characteristics file since it affects how the Matlab code functions, has no bearing on how any Python code functions (and if it does it should be isolated to that specific code base), and seems to only be causing unnecessary confusion. From another perspective, if the Matlab code changes how it interprets a limit, the only changes necessary should be in Matlab code.

I am happy to make any changes myself, but would like to come to a conclusion on Matlab identifiers first.

File: chandra_models/xija/dpa/dpa_spec.json
Comment (jzuhone) [reply chandra_models/xija/dpa/dpa_spec.json:270]: @matthewdahmer it's essentially the same kind of thing as the FP limits which are only valid in certain circumstances

**Comment by taldcroft** [comment]
> Why are we defining Matlab identifiers in the model json files? 

This was intended as a convenience for @jskrist and also to be sure that we are all on the same page since there is  ambiguity. As I was translating the MATLAB characteristics to this labeling nomenclature, I was having to think about things, so realized that @jskrist might have to spend time inverting what I did.

I originally wanted to make these as comments but JSON does not support comments. However I realized after the fact that the `matlab_identifier` labels could be used to programmatically set the MATLAB characteristics so there seemed to be some value-added.

Having said that I am not attached to these labels.

**Comment by taldcroft** [comment]
> When to include limits
As is currently the case, only those limits that are used should be included. This reduces required model maintenance, so these models do not need to be updated when unused limits are changed.

It is a little tricky to define limits being "used". Did you have something in mind? These `limits` reflect values that were defined in the MATLAB characteristics, so from my perspective they are all being used.

**Comment by taldcroft** [comment]
About deleting unused models, I'm fine with that. The intent of this PR was just to blindly transcribe the MATLAB characteristics limits to applicate xija models here. If at a later time models get deleted from here then no problem as long as it doesn't break MATLAB.

**Comment by matthewdahmer** [comment]
Regarding Matlab identifiers, if it really would be useful to @jskrist to provide a mapping that translates the limits included in each model with their corresponding Matlab counterparts, then perhaps this could be included in a single separate file for all these limits? 

I found when reviewing the limits that the matlab identifiers made it much more difficult for my brain to parse the limits sections. For example:
`"planning.warning.high.matlab_identifier": "yellow_limit"` doesn't really make sense to me due to different naming conventions colliding. Perhaps there is a good reason for this mapping, however there is no way to explain this in the json file (as you point out the JSON format does not support comments). A separate file that can support comments may be a better option, or perhaps it would be best to leave it to the Matlab code to interpret these limits and document the reasoning behind the naming conventions used if not obvious in the Matlab code itself.

Regarding when to use limits, I was just explicitly stating my preference that we not expand the use of the limit section beyond including the limits that are used in either the FOT Tools or in any of the ACIS tools for plotting, processing, etc. Anotherwords, there's no need to capture all the ODB limits for all MSIDs used in a given model. I may change my mind on this one eventually, but for now I'd like to keep this simple. I agree with the scope you used to define the `odb` and `planning` limits.

**Comment by taldcroft** [comment]
I don't feel strongly about the MATLAB identifiers, so they can be removed if @jskrist does not want them there for dynamically setting characteristics.

**Comment by jskrist** [comment]
While I like the idea of dynamically setting limits based on these identifiers, the biggest benefit I see to having the identifiers in the model spec file is the ability to add/change limits without a MATLAB Tools release.  If we don't want to be able to add limits or change a limit's type without a MATLAB Tools release, then I think this identifier mapping should live in the MATLAB Tools.

**Comment by taldcroft** [comment]
I think that changing limit types or adding limits would be outside the scope of "allowed" model spec changes without a MATLAB release. I.e. the updates are intended to address changes in numerical values, but not the structure of the data.

So I'm happy enough to remove all those MATLAB identifiers. They'll still be available for reference in 8fffc8f.

**Comment by jzuhone** [comment]
I'm with @matthewdahmer on the fact that the matlab identifiers sort of clutter up the structure for me.

**Comment by taldcroft** [comment]
The MATLAB identifiers are gone!

**Comment by jskrist** [comment]
Where did we land wrt the updating the tcylaft6 spec file vs the minusyz spec?

**Comment by matthewdahmer** [comment]
I've been thinking about this some more, and one of the things that I realized was bothering me about the Matlab identifiers is that the mapping should be able to be inferred from the `planning` and `odb` limits themselves, and it seems to me the only reason the explicit mapping arose in the first place is that the mapping on the Matlab side is not consistent (not pointing fingers, I understand how this happens). If it were made to be consistent, and the code were in place on the Matlab side to interpret the `planning` and `odb` limits, then we _could_ add limits to a model spec file and have them appear in MCC without a new release. The mappings below seem to be the most correct way to translate these limits, with the understanding that there can be multiple versions of each limit (e.g. multiple red limits), as there could be low and high versions. If this were to cause issues, then the `_low` and `_high` designation could be added to these designations on the Matlab side.
```
`reference_limit` = `planning.warning.<low, high>`
`penalty_limit` = `planning.penalty.<low, high>`
`red_limit` = `odb.warning.<low, high>`
`yellow_limit` = `odb.caution.<low, high>`
```
I understand my suggestion shifts the burden to @jskrist to modify the Matlab code, however this would standardize the Matlab usage of these limits, making the code easier to read and understand. Please let me know if I've misunderstood something.

**Comment by matthewdahmer** [comment]
@jskrist since the tcylaft6 and minusyz models are not used to limit temperatures anymore, they don't need limits, however if they did have them defined in their model spec files, then that would be fine still. It doesn't matter since they will be removed in a future release.

**Comment by taldcroft** [comment]
I just pushed a commit to add limits to `tcylaft6_spec.json`. The scope of this PR is to put limits corresponding to every single MSID in the MATLAB thermal characteristics, without respect to whether they are used operationally. If as @matthewdahmer says we later remove those models from MATLAB and chandra_models, then that's fine, but out of scope here.

**Comment by taldcroft** [comment]
@matthewdahmer - all good to try to clean things up, but I am also very much in favor of small incremental changes to not derail progress in the face of many competing priorities. Changing two things at once may not be a good idea. I think we're almost there with this change and so I'd propose tabling the idea of a bigger rework which does not actually have much impact.

**Comment by matthewdahmer** [comment]
@taldcroft I am 100% in agreement with your perspective:

> @matthewdahmer - all good to try to clean things up, but I am also very much in favor of small incremental changes to not derail progress in the face of many competing priorities. Changing two things at once may not be a good idea. I think we're almost there with this change and so I'd propose tabling the idea of a bigger rework which does not actually have much impact.

Removing models and considering further changes on the Matlab should be separate from both this PR and the upcoming Matlab release, and completed when the time is right.

**Comment by jzuhone** [comment]
@taldcroft the `limits` dictionary for the ACIS FP should look like this:

```json
    "limits": {
        "fptemp": {
            "planning.data_quality.high.acisi": -112.0,
            "planning.data_quality.high.aciss": -111.0,
            "planning.data_quality.high.aciss_hot": -109.0,
            "planning.data_quality.high.cold_ecs": -118.2,
            "planning.warning.high": -84.0,
            "odb.caution.high": -80.0,
            "unit": "degC"

        }
    },

```
The only problem is that `FPTEMP_11` is not in the ODB so it's not correct to refer to the -80 C limit as such, but I put it there as a placeholder for a better name.

**Review by matthewdahmer — APPROVED** [review]

**Comment by matthewdahmer** [comment]
All, I made the four changes I pointed out as necessary, updated the version in `__init__.py`, and added the ACIS FP limits.

@jzuhone I used `safety.caution.high` instead `odb.caution.high` since there isn't a record of `fptemp_11` in the ODB. I went back and forth on this one, but using a different category seems to be slightly less confusing.

**Review by jzuhone — APPROVED** [review]

**Comment by taldcroft** [comment]
@matthewdahmer - just to double-check, in your review did you confirm that each of the numbers match your expectation from the MATLAB thermal characteristics? There is a possibility of copy/paste error (I did catch one earlier myself).

**Comment by matthewdahmer** [comment]
@taldcroft I did but I paid more attention to the planning limits than the others, just found one OBA ODB limit that needed to be adjusted.

**Comment by matthewdahmer** [comment]
I'll add that the limits in these JSON files were checked to ensure they reflect the guideline and ODB limits, so it is possible one or more ODB limits in these JSON files may not match what is currently in the MATLAB characteristics (e.g. `Thermal.msid_4rt700t.red_limit `), however this is intended.

**Comment by taldcroft** [comment]
@matthewdahmer - good catch, that's what I was worried about. I can't click on the Approve button for my own PR, but this looks good to me!

**Pull request merged**
